torch/csrc/dynamo/python_compiled_autograd.cpp

```diff
@@ -0,0 +1,479 @@
+    _methods};
+
+static TraceState call_begin_capture(
+    PyObject* self,
+    CacheNode& cache,
+    AutogradCompilerCall& compiler_call,
+    size_t num_outputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("begin_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(compiler_call.inputs));
+  THPObjectPtr pysizeinput(cache.wrap_dynamic_inputs());
+  THPObjectPtr pyresult(check(PyObject_CallMethodObjArgs(
+      self, method_name, pyinput.get(), pysizeinput.get(), NULL)));
+
+  PyObject *fake_inputs, *fake_sizes;
+  check(PyArg_ParseTuple(pyresult.get(), "OO", &fake_inputs, &fake_sizes));
+  return TraceState(
+      THPVariable_UnpackList(fake_inputs),
+      cache.unwrap_dynamic_inputs(fake_sizes),
+      compiler_call.accumulate_grad,
+      num_outputs);
+}
+
+static PyObject* call_end_capture(PyObject* self, const variable_list& inputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("end_capture");
+  THPObjectPtr pyinput(THPVariable_WrapList(inputs));
+  return check(PyObject_CallMethodOneArg(self, method_name, pyinput.get()));
+}
+
+variable_list compiled_autograd(
+    const std::shared_ptr<Node>& graph_root,
+    GraphTask& graph_task,
+    bool accumulate_grad,
+    const edge_list& output_edges) {
+  TORCH_CHECK(
+      output_edges.empty() || !accumulate_grad,
+      "outputs+accumulate_grad not yet implemented")
+  static std::mutex lock;
+  std::lock_guard<std::mutex> lock_guard(lock);
+  pybind11::gil_scoped_acquire gil;
+  std::unordered_map<Node*, int> dependencies =
+      std::move(graph_task.dependencies_);
+  std::vector<std::shared_ptr<Node>> worklist{graph_root};
+  worklist.reserve(dependencies.size());
+  AutogradCompilerCall compiler_call(accumulate_grad);
+  for (const auto i : c10::irange(output_edges.size())) {
+    compiler_call.node_calls.lookup(output_edges[i].function)
+        .mark_output(output_edges[i].input_nr, i);
+  }
+  const bool check_exec_info = !graph_task.exec_info_.empty();
+  CacheNode* cache = CacheNode::root();
+  std::vector<NodeCall*> calls;
+  calls.reserve(dependencies.size() + 1);
+
+  while (!worklist.empty()) {
+    std::shared_ptr<Node> fn = std::move(worklist.back());
+    worklist.pop_back();
+    NodeCall& call = compiler_call.node_calls.lookup(fn);
+    calls.emplace_back(&call);
+
+    { // update cache and gather args into `compiler_call`
+      CompiledNodeArgs node_args(compiler_call, call);
+      fn->compiled_args(node_args);
+      node_args.collect(call);
+      cache = cache->lookup(node_args.key());
+    }
+
+    for (const auto& edge : fn->next_edges()) {
+      if (!edge.is_valid()) {
+        continue;
+      }
+      if (check_exec_info) {
+        auto it = graph_task.exec_info_.find(edge.function.get());
+        if (it == graph_task.exec_info_.end() || !it->second.should_execute()) {
+          continue;
+        }
+      }
+      auto it = dependencies.find(edge.function.get());
+      TORCH_CHECK(it != dependencies.end());
```

Comment:
nit: TORCH_CHECK -> TORCH_INTERNAL_ASSERT

Prefer to use TORCH_CHECK with a nice error message when the user is expected to reach this point, e.g .when they misuse the API. If its just to check an invariant in the code, prefer TORCH_INTERNAL_ASSERT.

Reply:
(might be worth double checking some of the other TORCH_CHECKs)